Changes to simplify final_suspend() semantics.

lewissbaker opened the pull request:
For paper P1477R0.

Commit:
Changes to simplify final_suspend() semantics.

For paper P1477R0.

GorNishanov commented on Coroutines/declarators.tex:
```diff
@@ -149,6 +149,13 @@
+The result of this expression must have type \tcode{std::experimental::coroutine_handle<}$Z$\tcode{>} for some type $Z$.
+The coroutine identified by the returned handle is resumed as if by calling the \tcode{resume()} method. 
+A subsequent call to \tcode{$h$.resume()} on the coroutine handle passed to \tcode{final_suspend()}
+results in undefined behaviour.
```
We already have wording in [coroutine.handle.resumption] that prohibits resuming at the final suspend point. There should not be two different normative statements defining the same thing.

GorNishanov commented:
Also need to update [await.expr] to strip out mention of final_suspend

lewissbaker commented on Coroutines/declarators.tex:
```diff
@@ -149,6 +149,13 @@
+When the coroutine reaches \textit{final suspend point} the coroutine is suspended and the expression \linebreak
+\tcode{$p$.final_suspend(std::experimental::coroutine_handle<$P$>::from_promise($p$))} is evaluated.
+The result of this expression must have type \tcode{std::experimental::coroutine_handle<}$Z$\tcode{>} for some type $Z$.
+The coroutine identified by the returned handle is resumed as if by calling the \tcode{resume()} method. 
```
@GorNishanov Does this section need some extra wording to require implementations to perform a tail-call here?

GorNishanov replied:
Yes, Grab the wording from [await.expr]

lewissbaker replied:
The wording in [await.expr] is just a note. Is that sufficient?
Are implementations allowed to ignore the note?

lewissbaker replied:
How does that look now?

GorNishanov replied:
Looks good.

Commit:
Remove reference to 'final suspend point' in [await.expr].

The final-suspend point is no longer defined in terms of co_await
so this exclusion is not required.

Commit:
Don't mention calling resume() as UB in description of final suspend point.

This is already covered by normative text in description of `coroutine_handle::resume()`.

Commit:
Note that resuming final_suspend() continuation should be a tail call.

lewissbaker commented on Coroutines/declarators.tex:
```diff
@@ -112,7 +112,7 @@
      try { @\textit{F}@ } catch(...) { @\textit{p}@.unhandled_exception(); }
   @\textit{final_suspend}@:
-     co_await @$p$@.final_suspend(); // final suspend point
+     // final suspend point
```
Does this still imply that the promise object and frame would be destructed if an exception was thrown from call to final_suspend()? Do we need to add some more wording to clarify that this is not the case?

GorNishanov replied:
Fixed by: http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p1356r0.html
